Applications/Spire/Source/Ui/ScrollArea.cpp

```diff
@@ -0,0 +1,165 @@
+    : QScrollArea(parent) {
+  setMouseTracking(true);
+  horizontalScrollBar()->setContextMenuPolicy(Qt::NoContextMenu);
+  verticalScrollBar()->setContextMenuPolicy(Qt::NoContextMenu);
+  setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+  setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+  m_h_scroll_bar_timer.setInterval(SCROLL_BAR_HIDE_TIME_MS);
+  connect(&m_h_scroll_bar_timer, &QTimer::timeout, this,
+    &ScrollArea::hide_horizontal_scroll_bar);
+  m_v_scroll_bar_timer.setInterval(SCROLL_BAR_HIDE_TIME_MS);
+  connect(&m_v_scroll_bar_timer, &QTimer::timeout, this,
+    &ScrollArea::hide_vertical_scroll_bar);
+}
+
+void ScrollArea::setWidget(QWidget* widget) {
+  widget->setMouseTracking(true);
+  widget->setAttribute(Qt::WA_Hover);
+  widget->installEventFilter(this);
+  QScrollArea::setWidget(widget);
+}
+
+bool ScrollArea::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::HoverMove) {
+    auto e = static_cast<QHoverEvent*>(event);
+    if(is_within_horizontal_scroll_bar(e->pos().y()) &&
+        !verticalScrollBar()->isVisible()) {
+      set_scroll_bar_style(SCROLL_BAR_MAX_SIZE);
+      setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+      setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    } else if(is_within_vertical_scroll_bar(e->pos().x()) &&
+        !horizontalScrollBar()->isSliderDown()) {
+      set_scroll_bar_style(SCROLL_BAR_MAX_SIZE);
+      setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOn);
+      setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    } else {
+      set_scroll_bar_style(SCROLL_BAR_MIN_SIZE);
+      if(!m_v_scroll_bar_timer.isActive() &&
+          verticalScrollBarPolicy() != Qt::ScrollBarAlwaysOff &&
+          !verticalScrollBar()->isSliderDown()) {
+        hide_vertical_scroll_bar();
+      }
+      if(!m_h_scroll_bar_timer.isActive() &&
+          horizontalScrollBarPolicy() != Qt::ScrollBarAlwaysOff &&
+          !horizontalScrollBar()->isSliderDown()) {
+        hide_horizontal_scroll_bar();
+      }
+    }
+  } else if(event->type() == QEvent::HoverLeave) {
+    if(!m_h_scroll_bar_timer.isActive() &&
+        horizontalScrollBarPolicy() == Qt::ScrollBarAlwaysOff) {
+      hide_horizontal_scroll_bar();
+    }
+    if(!m_v_scroll_bar_timer.isActive() &&
+        verticalScrollBarPolicy() == Qt::ScrollBarAlwaysOff) {
+      hide_vertical_scroll_bar();
+    }
+  }
+  return QScrollArea::eventFilter(watched, event);
+}
+
+void ScrollArea::leaveEvent(QEvent* event) {
+  hide_horizontal_scroll_bar();
+  hide_vertical_scroll_bar();
+}
+
+void ScrollArea::wheelEvent(QWheelEvent* event) {
+  if(event->modifiers() & Qt::ShiftModifier) {
+    setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+    horizontalScrollBar()->setValue(horizontalScrollBar()->value() -
+      (event->delta() / 2));
+    m_h_scroll_bar_timer.start();
+  } else if(!event->modifiers().testFlag(Qt::ShiftModifier)) {
```

Comment:
Is `event->modifiers() & Qt::ShiftModifier` any different from `event->modifiers().testFlag(Qt::ShiftModifier)`? API-wise, they seem to do the same thing. In case they do, you don't need the `else` condition and can move `m_h_scroll_bar_timer.start();` out.

Reply:
I can change it like this:

```
  if(event->modifiers().testFlag(Qt::ShiftModifier)) {
    ...scroll horizontally...
  } else {
    ...scroll vertically...
  }
```

But each condition does something different so I can't remove anything else.

Reply:
Oh, `h_` looks like "handle" from my point, so I didn't notice those are 2 different timers.
This comment mostly referred to the unnecessary check in the `else` condition itself.
Can you, however, get rid of those abbreviations?